devfile/devworkspace-operator
# Add container security context field to DevWorkspace Operator Config

### What does this PR do?
* Add `.config.workspace.containerSecurityContext` field to DevWorkspaceOperatorConfig to allow configuring container-level security context for workspace pods
* Update how podSecurityContext is handled to allow specifying a configured value in OpenShift as well (previously it was ignored)
* Merge podSecurityContext values into default configuration rather than overwriting it entirely
* Minor fixes:
  * Add config package test that fuzzes config to ensure all fields are considered by merging
  * Update CRD documentation
  * Allow specifying different defaults for OpenShift and Kubernetes for podSecurityContext
  * Log when a non-default container/pod security context is used

### What issues does this PR fix or reference?
Closes #951 
Closes #952 

### Is it tested? How?
To test:
* Verify that functionality is unchanged on OpenShift when no values are configured
* Verify that it's possible to configure both container and pod security context
* Verify that pod security context does not overwrite unset fields

On Kubernetes, the following DevWorkspaceOperatorConfig fields can be used to test:
```yaml
config:
  workspace:
    containerSecurityContext:
      capabilities:
        add:
        - SETUID
        - SETGID
    podSecurityContext:
      runAsUser: 2048
```
This should result in the workspace pod having `runAsUser: 2048` and pod containers having the `SETUID` and `SETGID` capabilities.

On OpenShift, the above DWOC will likely cause the workspace to fail to start. One permissible configuration is to specify a `runAsUser` within the UID range assigned the current project, e.g.
```yaml
config:
  workspace:
    containerSecurityContext:
      runAsUser: 1000670002
    podSecurityContext:
      runAsUser: 1000670001
```
(where `1000670000` is the default UID used for running workspace pods).

### PR Checklist
